helper/util/restart_linux.go

```diff
@@ -69,8 +70,45 @@ func (*containerRestarter) RestartContainer() error {
 	}
 
 	// kill the process group
+	path := "/proc/" + strconv.Itoa(pgid)
+
+	stderrlog.Infof("Sending SIGINT...")
 	_ = syscall.Kill(-pgid, syscall.SIGINT)
-	time.Sleep(2000)
+	
+	for i := 0; i < 5; i++ {
+		time.Sleep(time.Second)
+		_, err := os.Stat(path)
+		if os.IsNotExist(err) {
+			stderrlog.Infof("Process terminated!")
+			return nil;
+		}
+	}
```

Comment:
Could  use `wait.PollImmediate` here instead of a loop. It's already [included in the helper it seems](https://github.com/devspace-sh/devspace/blob/main/helper/util/pingtimeout/ping_timeout.go#L33).

Reply:
Updated. According to the [docs](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollImmediate), `wait.PollImmediate` is deprecated, but the version already included in the helper does not have the replacement function yet, so i used it anyway. Would it be better to update the import?

Alternatively, we could wait for the process using the [pidfd](https://github.com/oraoto/go-pidfd) interface. Supposedly this would be slightly more efficient, but also, more importantly, it would avoid a race condition in case the pid would be reused between sending the signal and polling for termination. It would require an extra import and might make the helper somewhat less portable though (the syscall appeared in Linux 5.3).

```go
func KillPidFd(pid int) {

        fmt.Printf("Opening process %d\n", pid)

        fd, err := pidfd.Open(pid, 0)
        if err != nil {
                fmt.Printf("Error: %s\n", err.Error())
                return
        }

        for _, sig := range []syscall.Signal{syscall.SIGINT, syscall.SIGTERM, syscall.SIGKILL} {
                fmt.Printf("Sending %s signal...\n", sig.String())

                err = fd.SendSignal(sig, 0)
                if err != nil {
                        fmt.Printf("Error: %s\n", err.Error())
                        return
                }

                fds := []unix.PollFd{{Fd: int32(fd), Events: unix.POLLIN}}
                n, _ := unix.Poll(fds, 5000 /* milliseconds */)

                if n > 0 {
                        fmt.Printf("Process terminated!\n")
                        return
                }
        }

        fmt.Printf("Timeout!\n")

}
```

Reply:
@maartendeprez 

This is already a good improvement over the original code. I think we can wait to upgrade to the replacement for `wait.PollImmediate`, so that it can upgraded outside the helper as well.

Regarding the `pidfd` interface, if the race condition has been observed in your testing, and happens frequently enough, then we can consider adding it. If it hasn't been observed then we can skip for now.